File: python_mid_june_2024/14. Longest Common Prefix.py
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        def compare(str1: str, str2:str) -> str:
            i, j = 0, 0
            m, n = len(str1), len(str2)
            ans = ""
            while i < m and j < n:
                if str1[i] != str2[j]:
                    break
                ans += str1[i]
                i += 1
                j += 1
            return ans
        
        common = strs[0]
        length = len(strs)
        for s in strs:
            common = compare(common, s)
        return common



        

File: java_july/14. Longest Common Prefix.java
class Solution {
    public String longestCommonPrefix(String[] strs) {
        int len = strs[0].length();
        int n = strs.length;
        String shortest = strs[0];
        for (String str : strs) {
            if (str.length() < len){
                shortest = str;
                len = str.length();
            }
        }
        if (len < 1 || shortest == "") {
            return "";
        }
        for (int i=0; i<len; i++) {
            char c = shortest.charAt(i);
            for (int j=0; j<n; j++) {
                if (strs[j].charAt(i) != c) {
                    return shortest.substring(0, i);
                }
            }
        }
        return shortest;
    }
}